tests/unit/test_tasks.py

```diff
@@ -225,7 +226,18 @@ def test_cleanups_only_ran_once_on_exception(self):
         for done_callback in done_callbacks:
             self.transfer_coordinator.add_done_callback(done_callback)
 
-        submission_task()
+        try:
+            submission_task()
+        except InvalidStateTransitionError:
```

Comment:
Do you think we should be catching it [here](https://github.com/boto/s3transfer/blob/develop/s3transfer/tasks.py#L128) instead? I did not do that is a good idea because I thought it would be worse to be covering up invalid state transitions.

The real problem lies in that a single task cannot guarantee that if it tries to set the result/exception that another thread has not come in and set a result and set the done event. To do that we need to couple state setting and done event setting under a single threading primitive, which would be part of the state revamp.

Do you have any ideas?